src/server/setupWebServer.ts

```diff
@@ -60,10 +63,89 @@ export async function setupWebServer(
   }
 
   // (2) Add an endpoint to upload files
+  const filenameFor = (originalName: string) =>
+    `${nanoid()}${path.extname(originalName)}`;
   const storage = multer.diskStorage({
     destination: uploadedFilesDir,
-    filename: (req, file, cb) =>
-      cb(null, `${nanoid()}${path.extname(file.originalname)}`),
+    filename: (req, file, cb) => cb(null, filenameFor(file.originalname)),
+  });
+
+  const handleUploadedFile = async (
+    file: {
+      path: string;
+      originalname: string;
+      filename: string;
+    },
+    allowedFileTypes: string
+  ) => {
+    const mimeType = await getMimeType(file.path);
+    if (mimeType === undefined) {
+      throw new Error("MimeType could not be determined for uploaded file.");
+    }
+
+    const isImage = isMimeTypeImage(mimeType);
+    const isAudio = isMimeTypeAudio(mimeType);
+
+    if (
+      allowedFileTypes !== "*" &&
+      ((allowedFileTypes === "image" && !isImage) ||
+        (allowedFileTypes === "audio" && !isAudio))
+    ) {
+      throw new Error(
+        `A file with mime type ${mimeType} cannot be uploaded as ${allowedFileTypes}.`
+      );
+    }
+
+    return {
+      originalFilename: file.originalname,
+      filename: file.filename,
+      mimeType,
+      ...(isImage
+        ? {
+            type: "image" as const,
+            ...(await getImageDimensions(file.path)),
+            blurHash: await calculateBlurHash(file.path),
+          }
+        : isAudio
+        ? {
+            type: "audio" as const,
+            duration: await getAudioDuration(file.path),
+          }
+        : { type: "other" as const }),
+    };
+  };
+
+  const downloadFile = (url: string, name: string): Promise<string> =>
+    new Promise((resolve, _reject) => {
+      const filepath = path.join(uploadedFilesDir, filenameFor(name));
+      request(url).pipe(
+        fs.createWriteStream(filepath).on("close", () => resolve(filepath))
+      );
+    });
+
+  app.post("/api/upload-remote", bodyParser.json(), async (req, res, next) => {
```

Comment:
This effectively turns Roc & Roll into a service to make HTTP GET requests to random URLs. There are two potential problems with that:
- people could use Roc & Roll as a proxy to DDOS others - we could add some rate limiting to at least reduce this risk
- people can use this endpoint to make requests to `localhost`, and the server's internal network. This could be dangerous depending on what else is running on the server and network.

Reply:
fair points, at least a DDOS of our local server was already pretty easy using the other endpoint though.

I suppose the "proper way" to address this would be to add a layer of authentication. Blocking calls into the local network should be necessary either way, though. Not sure if there's an elegant way to do that though, besides maintaining a whitelist.

Reply:
Another thing to watch out for: `file://` urls (`request` does not support them, but whatever we might use to replace it with might).

> fair points, at least a DDOS of our local server was already pretty easy using the other endpoint though.

Absolutely true - I am mostly worried about other servers.

Reply:
> Blocking calls into the local network should be necessary either way, though. Not sure if there's an elegant way to do that though, besides maintaining a whitelist.

A (more long-term) solution could be to use a separate, properly firewalled server or container to download stuff, or even something like AWS Lambda.